security/advancedtls/pemfile_provider.go

```diff
@@ -0,0 +1,179 @@
+/*
+ *
+ * Copyright 2020 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package advancedtls
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io/ioutil"
+	"time"
+
+	"google.golang.org/grpc/credentials/tls/certprovider"
+	"google.golang.org/grpc/grpclog"
+)
+
+var defaultIdentityInterval = 1 * time.Hour
+var defaultRootInterval = 2 * time.Hour
+
+var readKeyCertPairFunc = tls.LoadX509KeyPair
+
+var readTrustCertFunc = func(trustFile string) (*x509.CertPool, error) {
+	trustData, err := ioutil.ReadFile(trustFile)
+	if err != nil {
+		return nil, err
+	}
+	trustPool := x509.NewCertPool()
+	ok := trustPool.AppendCertsFromPEM(trustData)
+	if !ok {
+		return nil, fmt.Errorf("failed to call AppendCertsFromPEM")
+	}
+	return trustPool, nil
+}
+
+var logger = grpclog.Component("advancedtls")
+
+// PEMFileProviderOptions contains options to configure a PEMFileProvider.
+// Note that these fields will only take effect in construction time. Once the PEMFileProvider starts, changing fields in PEMFileProviderOptions will do nothing.
+type PEMFileProviderOptions struct {
+	// CertFile is the file path that holds identity certificate whose updates will be captured by a watching goroutine.
+	// Optional. If this is set, KeyFile must also be set.
+	CertFile string
+	// KeyFile is the file path that holds identity private key whose updates will be captured by a watching goroutine.
+	// Optional. If this is set, CertFile must also be set.
+	KeyFile string
+	// TrustFile is the file path that holds trust certificate whose updates will be captured by a watching goroutine.
+	// Optional.
+	TrustFile string
+	// IdentityInterval is the time duration between two credential update checks for identity certs.
+	// Optional. If not set, we will use the default interval.
+	IdentityInterval *time.Duration
+	// RootInterval is the time duration between two credential update checks for root certs.
+	// Optional. If not set, we will use the default interval.
+	RootInterval *time.Duration
+}
+
+// PEMFileProvider implements certprovider.Provider.
+// It provides the most up-to-date identity private key-cert pairs and/or root certificates.
+type PEMFileProvider struct {
+	identityDistributor *certprovider.Distributor
+	rootDistributor     *certprovider.Distributor
+	cancel              context.CancelFunc
+}
+
+// NewPEMFileProvider uses PEMFileProviderOptions to construct a PEMFileProvider.
```

Comment:
Maybe
`NewPEMFileProvider returns a new PEMFileProvider constructed using the provided options`